### PR TITLE
Step down as master when configured out of voting configuration

### PR DESCRIPTION
**Pull request by ywelsch**
Abdicates to another master-eligible node once the active master is reconfigured out of the voting configuration, for example through the use of voting configuration exclusions.

Follow-up to #37712

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java [hunk 1]
```diff
@@ -178,6 +178,7 @@ public void testSimpleClusterStatePublishing() throws InterruptedException {
             discoveryNodes, singleNodeConfig, singleNodeConfig, 42L), ackListener, Collections.emptySet());
 
         assertThat(publication.pendingPublications.keySet(), equalTo(discoNodes));
+        assertThat(publication.completedNodes().size(), equalTo(0));
```
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java]: I prefer using the `empty()` matcher so if it's not empty we get to see what was in it.

**Review by DaveCTurner — APPROVED** [review]
LGTM

**Review by DaveCTurner — CHANGES_REQUESTED** [review]
Actually, I changed my mind, I think this behaviour should be noted in the docs for the voting config exclusions API.

**Review by DaveCTurner — APPROVED** [review]
LGTM

**Comment by ywelsch** [comment]
This required more changes, I'll need another review.

**Review by DaveCTurner — CHANGES_REQUESTED** [review]
I asked for more harmony in how we prioritise nodes.

File: server/src/main/java/org/elasticsearch/cluster/coordination/Reconfigurator.java [hunk 2]
```diff
@@ -162,4 +167,12 @@ public VotingConfiguration reconfigure(Set<DiscoveryNode> liveNodes, Set<String>
             return currentConfig;
         }
     }
+
+    private TreeSet<String> masterFirstTreeSet(Collection<? extends String> items, DiscoveryNode masterNode) {
```
[the remark this answers is not in the document]
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/cluster/coordination/Reconfigurator.java]: I've pushed 4e5886c4f9f

**Comment by ywelsch** [comment]
@elasticmachine retest this please

**Review by DaveCTurner — APPROVED** [review]
LGTM

**Comment by ywelsch** [comment]
@elasticmachine run elasticsearch-ci/default-distro

**Pull request merged**